### PR TITLE
👨‍🏭 keep user write metadata as much as possible

### PR DESCRIPTION
**Pull request by OsmondJiang**
remove the limitation of converting input metadata to strong model:
- only read what we want to read
- don't touch any thing we don't want to write back
- make less changes
- author and title belongs to our output model, controlled by docfx

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4726)

### CODE REVIEW

**Comment by OsmondJiang** [comment]
#4703

**Comment by OsmondJiang** [comment]
diff are expected

File: src/docfx/build/page/BuildPage.cs [hunk 1]
```diff
@@ -94,9 +76,49 @@ internal static class BuildPage
             }
 
             return (errors, publishItem);
+
+            async Task<OutputModel> GenerateOutputMetadata()
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: static method.

File: src/docfx/build/metadata/MetadataProvider.cs [hunk 2]
```diff
@@ -91,7 +88,10 @@ public MetadataProvider(Docset docset, Cache cache)
 
             errors.AddRange(_schemaValidator.Validate(result));
 
-            return (errors, metadata);
+            var (validationErrors, metadataModel) = JsonUtility.ToObject<InputMetadata>(result);
+            errors.AddRange(validationErrors);
+
+            return (errors, result, metadataModel);
```
Comment (live1206) [inline src/docfx/build/metadata/MetadataProvider.cs]: so there is some content overlapping between `result` and `metadataModel`?

File: src/docfx/build/page/BuildPage.cs [hunk 3]
```diff
@@ -18,55 +18,37 @@ internal static class BuildPage
...
+            var (errors, isPage, outputModel, (inputMetadata, metadataObject)) = await Load(context, file);
+            var (generateErrors, outputMetadata) = await GenerateOutputMetadata(context, file, tocMap, inputMetadata, outputModel);
+            errors.AddRange(generateErrors);
+
+            var mergedOutputMetadata = new JObject();
+            JsonUtility.Merge(mergedOutputMetadata, metadataObject);
+            JsonUtility.Merge(mergedOutputMetadata, JsonUtility.ToJObject(outputMetadata));
+            var outputPath = file.GetOutputPath(outputMetadata.Monikers, file.Docset.SiteBasePath, isPage);
+
+            var output = (object)null;
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
            object output = null;
```

File: src/docfx/build/page/BuildPage.cs [hunk 4]
```diff
@@ -18,55 +18,37 @@ internal static class BuildPage
...
+            var metadata = (JObject)null;
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
            JObject metadata = null;
```

File: src/docfx/build/page/BuildPage.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/page/BuildPage.cs]: hmmm...

File: src/docfx/build/metadata/MetadataProvider.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/build/metadata/MetadataProvider.cs]: result will be only for reading

**Pull request merged**